os/arch/arm/src/amebasmart/amebasmart_idle.c

```diff
@@ -140,6 +140,7 @@ But in case that happens, we have to introduce some strategy to handle that
 								arm_cpu_sgi(GIC_IRQ_SGI4, (1 << 1));
 								arm_arch_timer_int_mask(0);
 								up_timer_enable();
+								pmdbg("Secondary core is still running!\n");
```

Comment:
At this point, we have already called pm_changestate and informed all drivers that the board is going into sleep. However, if the secondary core never goes to idle or if there is a delay in secondary core going to idle, then the system remains in previous power state, but the drivers assume sleep state. 

How about calling pm_changestate at line 160? In that case, we dont have to worry about the hw timer also.

Reply:
There are some changes in the PM structure, now we will use suspend/resume as according to each domain (ie. this current flow of pm_changestate() will be modified in other PR). Thus, it will not have the concept of whether each of the driver to assume any PM state. In return, if the secondary core still utilise any driver, it should be fine.
@ewoodev  Could you please share your opinion regarding this?

Reply:
I believe that if the board sleep is called as Kishore said, it must enter sleep mode. Otherwise, there would be no point in changing the state in the PM. At least, if the board sleep entry fails, it needs to be informed with a return error format.

And I am not sure if it is the right way to check the scheduling status of the second CPU in the BSP code.
If the second CPU is executing regular threads, the PM state should not become sleep either.

I suggest checking this in the pm_idle. We can check this by calling sched_cpu_pause. While the PM is trying to enter sleep, no CPU should be assigned a tick ISR. Scheduling should stop.

Reply:
If then, the logic for checking secondary core status should be moved out to pm_changestate() (ie, As I knew, this will decide whether we can do a state change?), and firing the SGI4 (ie. hotplug handler) to secondary core should be triggered in pm_idle().

Reply:
As I have checked, there seems to be no sched_cpu_pause in current codebase. Do you mean up_cpu_pause...?

Reply:
sorry,.. sched_pause_cpu()
i think, the logic for checking secondary core status should be moved out to pm_idle() only, and Can SGI4 stay up_pm_board_sleep()?

But please confirm if this is possible first.

Reply:
I have updated the logic, pm_idle() will handle checking all other core status according to CONFIG_SMP_NCPUS.
I don't know why we need to invoke sched_cpu_pause, could you explain the reason?
SGI4 can be done in up_pm_board_sleep(), but there is a slight time gap between "finish checking secondary core status"<->"send SGI in BSP". I'm afraid that the secondary core undergo context switch or some interrupt between this time period. (ie. There is no additional checking in hotplug handler now)

Reply:
the reason why we need to call sched_pause_cpu is also related to your concern.

Even if we check the scheduling status of the second cpu in pm_idle, the second cpu can receive tick ISR until SGI4 is called. If this is wrong, please correct it. It is thought that the scheduling status of the second cpu is allowed to change.

Reply:
Yes, I understand the pause is to freeze the status of secondary core. Then, the pause will be done in pm_idle. And as I understand, the resume need to be called so that the secondary core can handle SGI4, which means the resume will be done in amebasmart_idle.c, is that acceptable? Please correct if my understanding is wrong.